File: spikectl/ujsonrpc.py
from typing import Dict, Union
from base64 import b64decode
import json
import random
import string

RPC_KEY_ID = 'i'
RPC_KEY_METHOD = 'm'
RPC_KEY_RESULT = 'r'
RPC_KEY_ERROR = 'e'
RPC_KEY_PARAMETERS = 'p'

CR = '\r'


class RPCBaseMessage(object):
    def is_error(self): False

    def is_response(self): False

    def is_request(self): False

    def is_notification(self): False


class RPCNotification(RPCBaseMessage):

    __slots__ = ['method', 'parameters']

    def __init__(self, method: Union[str, int], parameters: any):
        self.method = method
        self.parameters = parameters

    def is_notification(self):
        return True

    def __str__(self):
        return 'RPCNotification [method: {}, parameters: {}]'.format(
            self.method,
            self.parameters
        )


class RPCError(RPCBaseMessage):

    __slots__ = ['id', 'exception']

    def __init__(self, idx: str, exception_data: str):
        self.id = idx
        decoded_exception_data = b64decode(exception_data)
        self.exception = str(decoded_exception_data, 'utf-8')

    def is_error(self):
        return True

    def __str__(self):
        return 'RPCError [id: {}, exception: {}]'.format(
            self.id,
            self.exception
        )


class RPCResponse(RPCBaseMessage):

    __slots__ = ['id', 'result']

    def __init__(self, idx: str, result: any):
        self.id = idx
        self.result = result

    def is_response(self):
        return True

    def __str__(self):
        return f'RPCResponse [id: {self.id}, result: {self.result}]'


def gen_idx() -> str:
    return ''.join(random.choices(string.ascii_letters + string.digits, k=4))


class RPCRequest(RPCBaseMessage):

    __slots__ = ['id', 'method', 'parameters']

    def __init__(self, method: str, parameters: any, idx: str):
        self.id = idx if idx else gen_idx()        
        self.method = method
        self.parameters = parameters

    def is_request(self):
        return True


def is_request(msg: any) -> bool:
    return RPC_KEY_ID in msg and RPC_KEY_METHOD in msg and RPC_KEY_PARAMETERS in msg


def is_response(msg) -> bool:
    return RPC_KEY_ID in msg and RPC_KEY_RESULT in msg


def is_error(msg) -> bool:
    return RPC_KEY_ID in msg and RPC_KEY_ERROR in msg


def is_notification(msg) -> bool:
    return RPC_KEY_METHOD in msg and RPC_KEY_PARAMETERS in msg and RPC_KEY_ID not in msg


def _to_json(obj: any) -> str:
    return json.dumps(obj, separators=(',', ':')) + CR


def encode_request(idx: str, method: str, parameters: any) -> str:
    return _to_json({
        RPC_KEY_ID: idx,
        RPC_KEY_METHOD: method,
        RPC_KEY_PARAMETERS: parameters
    })


def encode_response(idx: str, response: any) -> str:
    return _to_json({
        RPC_KEY_ID: idx,
        RPC_KEY_RESULT: response
    })


def encode_notification(method: str, parameters: any) -> str:
    return _to_json({
        RPC_KEY_METHOD: method,
        RPC_KEY_PARAMETERS: parameters
    })


def decode(msg: Dict[str, any]) -> RPCBaseMessage:
    if is_notification(msg):  # Check notifications first as those are more frequent
        return RPCNotification(msg[RPC_KEY_METHOD], msg[RPC_KEY_PARAMETERS])
    elif is_response(msg):
        return RPCResponse(msg[RPC_KEY_ID], msg[RPC_KEY_RESULT])
    elif is_error(msg):
        return RPCError(msg[RPC_KEY_ID], msg[RPC_KEY_ERROR])
    elif is_request(msg):
        return RPCRequest(msg[RPC_KEY_ID], msg[RPC_KEY_METHOD], msg[RPC_KEY_PARAMETERS])

def encode(msg: Union[RPCBaseMessage]) -> str:
    if isinstance(msg, RPCBaseMessage):
        if msg.is_request():
            request: RPCRequest = msg
            return encode_request(
                request.id,
                request.method,
                request.parameters
            )

    raise Exception(f'Unexcpected message type {type(msg)}')


__all__ = [
    'encode',
    'decode',
    'RPCBaseMessage',
    'RPCNotification',
    'RPCError',
    'RPCRequest',
    'RPCResponse'
]

File: spikedev-js/src/ujsonrpc.js
import { Transform } from 'stream'

const RPC_KEY_ID = 'i'
const RPC_KEY_METHOD = 'm'
const RPC_KEY_RESULT = 'r'
const RPC_KEY_ERROR = 'e'
const RPC_KEY_PARAMETERS = 'p'

export class RPCMessage {
    /**
     * @returns {boolean}
     */
    isError() {
        return false
    }
    /**
     * @returns {boolean}
     */
    isResponse() {
        return false
    }
    /**
     * @returns {boolean}
     */
    isRequest() {
        return false
    }
    /**
     * @returns {boolean}
     */
    isNotification() {
        return false
    }
}

export class RPCNotification extends RPCMessage {
    /**
     * 
         * @param {number|string} method 
         * @param {any} parameters 
         */
    constructor(method, parameters) {
        super()
        this.method = method
        this.parameters = parameters
    }

    isNotification() {
        return true
    }
}

export class RPCError extends RPCMessage {
    /**
     * @param {string} idx 
     * @param {string} exceptionData 
     */
    constructor(idx, exceptionData) {
        super()
        this.idx = idx
        let decoded = Buffer.from(exceptionData)
        this.exceptionData = decoded.toString('utf8')
    }

    isError() {
        return true
    }
}

export class RPCResponse extends RPCMessage {
    /**
     * @param {string} idx 
     * @param {any} result 
     */
    constructor(idx, result) {
        super()
        this.idx = idx
        this.result = result
    }

    isResponse() {
        return true
    }
}

/**
 * @returns {string}
 */
function genIdx() {
    return Math.random().toString(36).substring(2, 6)
}

export class RPCRequest extends RPCMessage {
    /**
     * 
     * @param {string} method 
     * @param {any} parameters 
     * @param {string?} idx 
     */
    constructor(method, parameters, idx = genIdx()) {
        super()
        this.idx = idx
        this.method = method
        this.parameters = parameters
    }

    isRequest() {
        return true
    }

    encode() {
        return JSON.stringify({
            RPC_KEY_ID: this.idx,
            RPC_KEY_METHOD: this.method,
            RPC_KEY_PARAMETERS: this.parameters
        })
    }
}

/**
 * 
 * @param {Object.<string, any>} msg
 * @returns {boolean}
 */
function isRequest(msg) {
    return RPC_KEY_ID in msg && RPC_KEY_METHOD in msg && RPC_KEY_PARAMETERS in msg
}

/**
 * 
 * @param {Object.<string, any>} msg
 * @returns {boolean}
 */
function isResponse(msg) {
    return RPC_KEY_ID in msg && RPC_KEY_RESULT in msg
}

/**
 * 
 * @param {Object.<string, any>} msg
 * @returns {boolean}
 */
function isError(msg) {
    return RPC_KEY_ID in msg && RPC_KEY_ERROR in msg
}

/**
 * 
 * @param {Object.<string, any>} msg
 * @returns {boolean}
 */
function isNotification(msg) {
    return RPC_KEY_METHOD in msg && RPC_KEY_PARAMETERS in msg && !(RPC_KEY_ID in msg)
}

/**
 * 
 * @param {Object.<string, any>} msg
 * @returns {RPCMessage}
 */
export function decode(msg) {
    if (isNotification(msg)) {
        const method = msg[RPC_KEY_METHOD]
        const parameters = msg[RPC_KEY_PARAMETERS]
        return new RPCNotification(method, parameters)
    } else if (isResponse(msg)) {
        const idx = msg[RPC_KEY_ID]
        const result = msg[RPC_KEY_RESULT]
        return new RPCResponse(idx, result)
    } else if (isError(msg)) {
        const idx = msg[RPC_KEY_ID]
        const error = msg[RPC_KEY_ERROR]
        return new RPCError(idx, error)
    } else if (isRequest(msg)) {
        const idx = msg[RPC_KEY_ID]
        const method = msg[RPC_KEY_METHOD]
        const parameters = msg[RPC_KEY_PARAMETERS]
        return new RPCRequest(idx, method, parameters)
    }
    throw new Error(`Unexcpected message type: ${msg}`)
}
